libyul/optimiser/UnusedStoreEliminator.cpp

```diff
@@ -0,0 +1,404 @@
+{
+	UnusedStoreBase::operator()(_functionCall);
+
+	for (Operation const& op: operationsFromFunctionCall(_functionCall))
+		applyOperation(op);
+
+	ControlFlowSideEffects sideEffects;
+	if (auto builtin = m_dialect.builtin(_functionCall.functionName.name))
+		sideEffects = builtin->controlFlowSideEffects;
+	else
+		sideEffects = m_controlFlowSideEffects.at(_functionCall.functionName.name);
+
+	if (!sideEffects.canContinue)
+	{
+		changeUndecidedTo(State::Unused, Location::Memory);
+		changeUndecidedTo(sideEffects.canTerminate ? State::Used : State::Unused, Location::Storage);
+	}
+}
+
+void UnusedStoreEliminator::operator()(FunctionDefinition const& _functionDefinition)
+{
+	ScopedSaveAndRestore storeOperations(m_storeOperations, {});
+	UnusedStoreBase::operator()(_functionDefinition);
+}
+
+
+void UnusedStoreEliminator::operator()(Leave const&)
+{
+	changeUndecidedTo(State::Used);
+}
+
+void UnusedStoreEliminator::visit(Statement const& _statement)
+{
+	using evmasm::Instruction;
+
+	UnusedStoreBase::visit(_statement);
+
+	auto const* exprStatement = get_if<ExpressionStatement>(&_statement);
+	if (!exprStatement)
+		return;
+
+	FunctionCall const* funCall = get_if<FunctionCall>(&exprStatement->expression);
+	if (!funCall)
+		return;
+	optional<Instruction> instruction = toEVMInstruction(m_dialect, funCall->functionName.name);
+	if (!instruction)
+		return;
+
+	if (!ranges::all_of(funCall->arguments, [](Expression const& _expr) -> bool {
+		return get_if<Identifier>(&_expr) || get_if<Literal>(&_expr);
+	}))
+		return;
+
+	// We determine if this is a store instruction without additional side-effects
+	// both by querying a combination of semantic information and by listing the instructions.
+	// This way the assert below should be triggered on any change.
+	using evmasm::SemanticInformation;
+	bool isStorageWrite = (*instruction == Instruction::SSTORE);
+	bool isMemoryWrite =
+		*instruction == Instruction::EXTCODECOPY ||
+		*instruction == Instruction::CODECOPY ||
+		*instruction == Instruction::CALLDATACOPY ||
+		*instruction == Instruction::RETURNDATACOPY ||
+		*instruction == Instruction::MSTORE ||
+		*instruction == Instruction::MSTORE8;
+	bool isWriteToBeHandled =
+		SemanticInformation::otherState(*instruction) != SemanticInformation::Write && (
+			SemanticInformation::storage(*instruction) == SemanticInformation::Write ||
+			(!m_ignoreMemory && SemanticInformation::memory(*instruction) == SemanticInformation::Write)
+		);
+	yulAssert(isWriteToBeHandled == (isStorageWrite || (!m_ignoreMemory && isMemoryWrite)));
+	if (isWriteToBeHandled)
+	{
+		m_stores[YulString{}].insert({&_statement, State::Undecided});
+		vector<Operation> operations = operationsFromFunctionCall(*funCall);
+		yulAssert(operations.size() == 1, "");
+		m_storeOperations[&_statement] = move(operations.front());
+	}
+}
+
+void UnusedStoreEliminator::finalizeFunctionDefinition(FunctionDefinition const&)
+{
+	changeUndecidedTo(State::Used);
+	scheduleUnusedForDeletion();
```

Comment:
Is this needed here?

Reply:
How would it work without that?

Reply:
Won't the ones that are unused at this point just remain unused and ``scheduleUnusedForDeletion()`` in
``UnusedStoreEliminator::run`` will just take care of them later in bulk?

Reply:
Ah, they'll be removed due to the scoped saving thing in the base class maybe?

Reply:
Yeah, they will... although we probably wouldn't really have to do that, would we? But even then, better to keep the map small. Yeah, better to just keep it, so nevermind!